### PR TITLE
Generic datastore service with basic indexing and querying capabilities, ts client generation

### PR DESCRIPTION
**Pull request by crufter**

### CODE REVIEW

File: datastore/proto/datastore.proto [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	// eg. age
+	string fieldName = 1;
+	// Ordered or unordered keys. Ordered keys are padded.
+	// Default is true. This option only exists for strings, where ordering
+	// comes at the cost of having rather long padded keys.
+	enum OrderType {
+    	option allow_alias = true;
+    	UNORDERED = 0;
+    	ASCENDING = 1;
+    	DESCENDING = 1;
```
Comment (domwong) [inline datastore/proto/datastore.proto:37]: `s/1/2/` ??

**Review by domwong — APPROVED** [review]
Overall looks good - most of the heavy lifting is done by the model stuff anyway. Would be good to get the README or some other documentation sorted.

File: datastore/handler/datastore.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+		indexes = append(indexes, v.Index)
+	}
+	return indexes, nil
+}
+
+func (e *Datastore) Read(ctx context.Context, req *datastore.ReadRequest, rsp *datastore.ReadResponse) error {
+	log.Info("Received Datastore.Read request")
+	q := toQuery(req.Query)
+	result := []map[string]interface{}{}
+	indexes, err := e.getIndexes(ctx)
```
[the remark this answers is not in the document]
Comment (crufter) [reply datastore/handler/datastore.go:81]: Yes your observation is very astute I could load the indexes at startup time and put it in memory as an optimization but I wanted to wait until we are sure this is the direction to take.

File: datastore/proto/datastore.proto (hunk 1 above)
Comment (crufter) [reply datastore/proto/datastore.proto:37]: oopsi

**Pull request merged**